Foundation/NSDecimal.swift

```diff
@@ -1,155 +1,1838 @@
+        if _isNegative == 1 {
+            return .negativeNormal
+        } else {
+            return .positiveNormal
+        }
+    }
+    public var isSignMinus: Bool {
+        return _isNegative != 0
+    }
+    public var isNormal: Bool {
+        return !isZero && !isInfinite && !isNaN
+    }
+    public var isFinite: Bool {
+        return !isNaN
+    }
+    public var isZero: Bool {
+        return _length == 0 && _isNegative == 0
+    }
+    public var isSubnormal: Bool {
+        return false
+    }
+    public var isInfinite: Bool {
+        return false
+    }
+    public var isNaN: Bool {
+        return _length == 0 && _isNegative == 1
+    }
+    public var isSignaling: Bool {
+        return false
+    }
 }
 
 extension Decimal : CustomStringConvertible {
     public init?(string: String, locale: Locale? = nil) {
-        // Avoid a compiler warning for now by not calling NSUnimplemented
-        return nil
+        let scan = Scanner(string: string)
+        var theDecimal = Decimal()
+        if !scan.scanDecimal(&theDecimal) {
+            return nil
+        }
+        self = theDecimal
+    }
+    public var description: String {
+        if self.isNaN {
+            return "NaN"
+        }
+        if _length == 0 {
+            return "0"
+        }
+        var copy = self
+        let ZERO : CChar = 0x30 // ASCII '0' == 0x30
+        let decimalChar : CChar = 0x2e // ASCII '.' == 0x2e
+        let MINUS : CChar = 0x2d // ASCII '-' == 0x2d
+
+        let bufferSize = 200 // max value : 39+128+sign+decimalpoint
+        var buffer = Array<CChar>(repeating: 0, count: bufferSize)
+
+        var i = bufferSize - 1
+        while copy._exponent > 0 {
+            i -= 1
+            buffer[i] = ZERO
+            copy._exponent -= 1
+        }
+
+        if copy._exponent == 0 {
+            copy._exponent = 1
+        }
+
+        while copy._length != 0 {
+            var remainder: UInt16 = 0
+            if copy._exponent == 0 {
+                i -= 1
+                buffer[i] = decimalChar
+            }
+            copy._exponent += 1
+            (remainder,_) = divideByShort(&copy, 10)
+            i -= 1
+            buffer[i] = Int8(remainder) + ZERO
+        }
+        if copy._exponent <= 0 {
+            while copy._exponent != 0 {
+                i -= 1
+                buffer[i] = ZERO
+                copy._exponent += 1
+            }
+            i -= 1
+            buffer[i] = decimalChar
+            i -= 1
+            buffer[i] = ZERO
+        }
+        if copy._isNegative != 0 {
+            i -= 1
+            buffer[i] = MINUS
+        }
+        return String(cString: Array(buffer.suffix(from:i)))
     }
-    public var description: String { NSUnimplemented() }
 }
 
-public func NSDecimalIsNotANumber(_ dcm: UnsafePointer<Decimal>) -> Bool { NSUnimplemented() }
+fileprivate func divideByShort<T:VariableLengthNumber>(_ d: inout T, _ divisor:UInt16) -> (UInt16,NSDecimalNumber.CalculationError) {
```

Comment:
When we added functionality to Decimal to have methods it became quite clear that it is nearly impossible to mimic the behavior and avoid needing to call out to the freestanding functions. They will most likely need to remain API in swift to service the hoisted method implementations. The divideByShort and fellow methods are relatively complex and hold some quirky behavior that these seem to attempt to mimic - we should consider porting some of the edge case tests from Foundation to swift-corelibs